_posts/2021-09-13-user-defined-value-types.md

```diff
@@ -0,0 +1,152 @@
+above abstraction has a runtime overhead, i.e., additional gas when compared to using just `uint128`
+to represent the underlying value. In particular, the functions `toPrice` and `toQuantity` involve
+storing the value in memory. Similarly, the functions `fromPrice` and `fromQuantity` read the respective
+value from memory. Together, these functions pass the value from `stack -> memory -> stack` which
+is often unnecessary and incurs a runtime cost. This issue is solved by user defined value types,
+which are abstractions of elementary value types (such as `uint8` or `address`), without any
+additional runtime overhead.
+
+## Syntax for User Defined Value Types
+
+A user defined value type is defined using `type C is V`, where `C` is the name of the newly
+introduced type and `V` has to be a built-in value type (the "underlying type"). The function
+`C.wrap` is used to convert from the underlying type to the custom type. Similarly, the function
+`C.unwrap` is used to convert from the custom type to the underlying type.
+
+Going back to the problem from the [motivation](#motivation) section, one can replace the structs by:
+
+```solidity
+pragma solidity ^0.8.8;
+
+type Price is uint128;
+type Quantity is uint128;
+```
+
+The functions `toPrice` and `toQuantity` can be replaced by `Price.wrap` and `Quantity.wrap`
+respectively. Similarly, the functions `fromPrice` and `fromQuantity` can be replaced by
+`Price.unwrap` and `Quantity.unwrap` respectively.
+
+The data-representation of the values of such types are inherited from the underlying type and the
+underlying type is also used in the ABI. This means that the following two `transfer` functions
+would be identical, i.e., they have the same [function
+selector](https://docs.soliditylang.org/en/latest/abi-spec.html#function-selector) as well as the
+same [ABI encoding and
+decoding](https://docs.soliditylang.org/en/latest/abi-spec.html#mapping-solidity-to-abi-types). This
+allows using user defined value types in a backwards compatible way.
+
+```solidity
+pragma solidity ^0.8.8;
+
+type Decimal18 is uint256;
+
+interface MinimalERC20 {
+    function transfer(address to, Decimal18 value) external;
+}
+
+interface AnotherMinimalERC20 {
+    function transfer(address to, uint256 value) external;
+}
+```
+
+Notice in the above example, how the user defined type `Decimal18` makes it clear that a value is
+supposed to represent a number with 18 decimals.
+
+## Example
+
+The following example illustrates a custom type `UFixed256x18` representing a decimal fixed point
+type with 18 decimals and a minimal library to do arithmetic operations on the type.
+
+```solidity
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.8;
+
+// Represent a 18 decimal, 256 bit wide fixed point type using a user defined value type.
+type UFixed256x18 is uint256;
+
+/// A minimal library to do fixed point operations on UFixed256x18.
+library FixedMath {
```

Comment:
Do we really need the library? There is no `using` statement.

Reply:
Good point. Maybe an average user might not care about a `library` example. On the other hand, a fixed point library is somewhat of a canonical use-case for user defined value types. Technically, you could just use `FixedMath.add(...)` instead of `UFixed256x18.wrap(5).add(...)`.